<a href="https://colab.research.google.com/github/daje0601/data_structure/blob/master/3_%EC%9E%AC%EA%B7%80%ED%95%A8%EC%88%98.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 3. 재귀함수

-. 자료구조 재귀함수에 대한 예제 코드 입니다.   
-. 재귀함수에 대한 자세한 설명은 아래 링크를 참조해 주세요  
-. https://daje0601.tistory.com/213


In [20]:
def count_down_func(n):
  if n > 0 :
    print(n)
    return count_down_func(n-1)
 
print(count_down_func(5))

5
4
3
2
1
None


In [18]:
def factorial_func(n):
    if n == 1:
    	return 1
    return n * factorial_func(n-1)

print(factorial_func(5))

120


간단히 설명을 드리면 처음에 n=5입니다. 

그럼 if절을 생각하고 return으로 넘어가게 됩니다. 

-> 5  * factorial_func(4)

-> 5  * 4 * factorial_func(3)

-> 5  * 4 * 3 * factorial_func(2)

-> 5  * 4 * 3 * 2 * factorial_func(1)

그러다가 factorial_func(1)은 1을 리턴해주기 때문에 1을 곱해주면 120이 출력이 됩니다. 

In [8]:
my_list = [1,2,3,4,5]
def sum_list(items):
    sum = 0
    for i in items:
        sum = sum + i
    return sum

sum_list(my_list)

15

In [17]:
def sum_list(items):
    if len(items) == 1:
        return items[0]
    else:
        return items[0] + sum_list(items[1:])  
sum_list(my_list)

15

위에 펙토리얼과 정말 똑같습니다. 단지 들어오는 값이 리스트로 변경되었다는 것뿐입니다.  
리스트의 len를 줄여주는 역할을 sum_list(items[1:])이 하게 됩니다.  
천천히 여유를 가지고 생각하시면 이해가 되실 겁니다. 